### PR TITLE
tls: request client certificate

### PR DESCRIPTION
**Pull request by marcosrmendezthd**
Description:

Allows Envoy to request a client certificate in cases where more than one type of client authentication is required. For example, some API gateways allow both mutual TLS (where a client certificate is needed) and OAuth2 (which uses a JWT). While this can be done with different listeners or a filter match on the SNI, this use case requires multiple types of authentication on the same SNI. This pull request adds the capability for Envoy to _request_, but not _require_ a client certificate. If the client provides a certificate, it will be validated against the configured downstream validation context.

Risk Level: Low

Testing:

Added unit tests to verify that `require_client_certificate` takes precedence over `request_client_certificate` and that the certificate is validated against the configured validation context.

Docs Changes:

Only the new option needs to be documented.

### CODE REVIEW

**Comment by junr03** [comment]
@marcosrmendezthd looks like DCO is failing. Do you mind taking a look?

**Comment by marcosrmendezthd** [comment]
> @marcosrmendezthd looks like DCO is failing. Do you mind taking a look?

yeah thought i fixed that...

**Comment by asraa** [comment]
It looks like some of the older commits are missing the sign-off, can you try the instructions here please? https://github.com/envoyproxy/envoy/pull/9914/checks?check_run_id=423656568

**Comment by marcosrmendezthd** [comment]
> It looks like some of the older commits are missing the sign-off, can you try the instructions here please? https://github.com/envoyproxy/envoy/pull/9914/checks?check_run_id=423656568

will do!

**Comment by marcosrmendezthd** [comment]
> It looks like some of the older commits are missing the sign-off, can you try the instructions here please? https://github.com/envoyproxy/envoy/pull/9914/checks?check_run_id=423656568

that worked thank you!!!

**Comment by asraa** [comment]
No problem! You'll also need to run the format script to fix the format error:
`./ci/run_envoy_docker.sh './ci/do_ci.sh fix_format'` with docker or `tools/check_format.py fix`

**Comment by marcosrmendezthd** [comment]
> No problem! You'll also need to run the format script to fix the format error:
> `./ci/run_envoy_docker.sh './ci/do_ci.sh fix_format'` with docker or `tools/check_format.py fix`

on it!

**Comment by marcosrmendezthd** [comment]
> Thank you for the tests! Would you be able to add quick line to document the added field in `docs/root/intro/version_history.rst`?

done

**Comment by marcosrmendezthd** [comment]
checking `//test/common/network:connection_impl_test ` locally.

**Comment by marcosrmendezthd** [comment]
test passed locally...

```
$ bazel test //test/common/network:connection_impl_test
...

INFO: From CcConfigureMakeRule external/envoy/bazel/foreign_cc/gperftools_build/include:
Target //test/common/network:connection_impl_test up-to-date:
  bazel-bin/test/common/network/connection_impl_test

INFO: Elapsed time: 837.272s, Critical Path: 260.09s
INFO: 1615 processes: 1615 darwin-sandbox.
INFO: Build completed successfully, 1657 total actions
//test/common/network:connection_impl_test                               PASSED in 4.0s

Executed 1 out of 1 test: 1 test passes.
INFO: Build completed successfully, 1657 total actions
```

**Comment by asraa** [comment]
Yep, it doesn't look related to your changes, so I re-ran the macOS check

**Comment by marcosrmendezthd** [comment]
> Yep, it doesn't look related to your changes, so I re-ran the macOS check

thank you! :D

File: docs/root/intro/version_history.rst [hunk 1]
```diff
@@ -16,7 +16,8 @@ Version history
   "envoy.reloadable_features.new_http2_connection_pool_behavior" and then re-configure your clusters or
   restart Envoy. The behavior will not switch until the connection pools are recreated. The new
   circuit breaker behavior is described :ref:`here <arch_overview_circuit_break>`.
-* upstream: changed load distribution algorithm when all priorities enter :ref:`panic mode<arch_overview_load_balancing_panic_threshold>`. 
+* upstream: changed load distribution algorithm when all priorities enter :ref:`panic mode<arch_overview_load_balancing_panic_threshold>`.
+* tls: added :ref:`request_client_certificate <envoy_api_field_auth.DownstreamTlsContext.request_client_certificate>` to request client certificates without requiring them.
```
Comment (lizan) [inline docs/root/intro/version_history.rst]: nit: alphabetical order

File: api/envoy/extensions/transport_sockets/tls/v3/cert.proto [hunk 2]
```diff
@@ -431,6 +431,9 @@ message DownstreamTlsContext {
   // certificate.
   google.protobuf.BoolValue require_client_certificate = 2;
 
+  // If specified, Envoy will request a client certificate.
+  google.protobuf.BoolValue request_client_certificate = 7;
```
Comment (lizan) [inline api/envoy/extensions/transport_sockets/tls/v3/cert.proto]: How does this play with `require_client_certificate`, can you add comment about that?

File: source/extensions/transport_sockets/tls/context_impl.cc [hunk 3]
```diff
@@ -989,6 +989,10 @@ ServerContextImpl::ServerContextImpl(Stats::Scope& scope,
       });
 
   for (auto& ctx : tls_contexts_) {
+    if (config.requestClientCertificate()) {
+      SSL_CTX_set_verify(ctx.ssl_ctx_.get(), SSL_VERIFY_PEER, nullptr);
```
Comment (lizan) [inline source/extensions/transport_sockets/tls/context_impl.cc]: Looks like we already do similar thing in https://github.com/envoyproxy/envoy/blob/master/source/extensions/transport_sockets/tls/context_impl.cc#L112-L122 (this was merged yesterday), so does that fill the actual use case of `request_client_certificate`? cc @jimini-lumox

Comment (marcosrmendezthd) [reply source/extensions/transport_sockets/tls/context_impl.cc]: going to test that out, maybe it does. 😄

Comment (PiotrSikora) [reply source/extensions/transport_sockets/tls/context_impl.cc]: This ignores and removes the flags already set in `ContextImpl::ContextImpl` (regardless of the yesterday's commit).

File: api/envoy/api/v2/auth/cert.proto [hunk 4]
```diff
@@ -413,6 +413,9 @@ message DownstreamTlsContext {
   // certificate.
   google.protobuf.BoolValue require_client_certificate = 2;
 
+  // If specified, Envoy will request a client certificate.
+  google.protobuf.BoolValue request_client_certificate = 7;
```
Comment (PiotrSikora) [inline api/envoy/api/v2/auth/cert.proto]: Certificate is always requested if `trusted_ca` is specified, so this is a no-op.

File: source/extensions/transport_sockets/tls/context_impl.cc (hunk 3 above)
Comment (marcosrmendezthd) [reply source/extensions/transport_sockets/tls/context_impl.cc]: @lizan sorry for the delay, had to test a couple of scenarios. you are correct, https://github.com/envoyproxy/envoy/pull/9172 handles this use case!

**Comment by marcosrmendezthd** [comment]
closing as https://github.com/envoyproxy/envoy/pull/9172 handles our use case very well, CA is optional as well. 😄

**Pull request closed without merging**